examples/matching_engine/src/lib.rs

```diff
@@ -0,0 +1,203 @@
+///
+/// Implementing fractional price is more complicate. One complicacy is that we do not know to serialize
+/// fractions so that the serialization is in the same order as the fractions. But that is only a
+/// technical problem. If we want to keep the serialization order we can limit ourselves to fractions
+/// of the form say x / 100000.
+/// The next problem is that if we do the fractions, then the order can only be filled partially. And
+/// in a mathematical way, Euclidean divisions have to be done. This might be the right way to address
+/// the problem and could be of interest to other developers.
+#[derive(
+    Clone, Copy, Debug, PartialEq, PartialOrd, Deserialize, Serialize, SimpleObject, InputObject,
+)]
+#[graphql(input_name = "PriceInput")]
+pub struct Price {
+    pub price: u64,
+}
+
+/// We use the custom serialization for the Price so that the order of the serialization
+/// corresponds to the order of the Prices.
+impl CustomSerialize for Price {
+    fn to_custom_bytes(&self) -> Result<Vec<u8>, ViewError> {
+        let mut short_key = bcs::to_bytes(&self.price)?;
+        short_key.reverse();
+        Ok(short_key)
+    }
+
+    fn from_custom_bytes(short_key: &[u8]) -> Result<Self, ViewError> {
+        let mut bytes = short_key.to_vec();
+        bytes.reverse();
+        let value = bcs::from_bytes(&bytes)?;
+        Ok(value)
+    }
+}
+
+impl Price {
+    pub fn revert(&self) -> Self {
+        Price {
+            price: u64::MAX - self.price,
+        }
+    }
+}
+
+pub fn product_price_amount(price: Price, count: Amount) -> Amount {
+    count.try_mul(price.price as u128).expect("product")
+}
+
+/// Transfer operation back to the owners
+#[derive(Clone)]
+pub struct Transfer {
+    /// Beneficiary of the transfer
+    pub owner: AccountOwner,
+    /// Amount being transferred
+    pub amount: Amount,
+    /// Index of the token being transferred (0 or 1)
+    pub token_idx: u32,
+}
+
+/// An identifier for a buy or sell order
+pub type OrderId = u64;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum ModifyCount {
+    All,
+    Partial(Amount),
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum OrderNature {
+    /// A bid for buying token 1 and paying in token 0
+    Bid,
+    /// An ask for selling token 1, to be paid in token 0
+    Ask,
+}
+
+scalar!(OrderNature);
+
+#[derive(Clone, Debug, Serialize, Deserialize, SimpleObject)]
+pub struct KeyBook {
+    /// The corresponding price
+    pub price: Price,
+    /// The nature of the order
+    pub nature: OrderNature,
+    /// The owner used for checks
+    pub owner: AccountOwner,
+}
+
+#[derive(Default, Clone, Debug, Serialize, Deserialize, SimpleObject)]
+pub struct AccountInfo {
+    /// The list of orders
+    pub orders: BTreeSet<OrderId>,
+}
+
+/// The order entry in the order book
+#[derive(Clone, Debug, Deserialize, Serialize, SimpleObject)]
+pub struct OrderEntry {
```

Comment:
Does this need to be part of the lib.rs? (which everyone includes to obtain the ABI)

Reply:
Ok, moved.